File: src/main/java/net/ccbluex/liquidbounce/ui/client/gui/clickgui/style/styles/onetap/Settings/TextSetting.java
package net.ccbluex.liquidbounce.ui.client.gui.clickgui.style.styles.onetap.Settings;

import net.ccbluex.liquidbounce.LiquidBounce;
import net.ccbluex.liquidbounce.ui.client.gui.ClickGUIModule;
import net.ccbluex.liquidbounce.features.module.modules.client.HUD;
import net.ccbluex.liquidbounce.ui.client.gui.clickgui.style.styles.onetap.*;
import net.ccbluex.liquidbounce.ui.client.gui.clickgui.style.styles.onetap.Settings.*;
import net.ccbluex.liquidbounce.ui.client.gui.clickgui.style.styles.onetap.Utils.Position;
import net.ccbluex.liquidbounce.ui.font.Fonts;
import net.ccbluex.liquidbounce.utils.render.RenderUtils;
import net.ccbluex.liquidbounce.features.value.TextValue;
import net.ccbluex.liquidbounce.features.value.Value;
import org.lwjgl.input.Keyboard;

import java.awt.*;

public class TextSetting extends Downward<TextValue>  {
    public TextValue textValue;
    private float modulex;
    private float moduley;
    private float texty;
    
    public TextSetting(final TextValue s, final float x, final float y, final int width, final int height, final ModuleRender moduleRender) {
        super(s, x, y, width, height, moduleRender);
        this.textValue = s;
    }
    
    @Override
    public void draw(final int mouseX, final int mouseY) {
        int guiColor = ClickGUIModule.generateColor().getRGB();
        this.modulex = OtcClickGUi.getMainx();
        this.moduley = OtcClickGUi.getMainy();
        this.texty = this.pos.y + this.getScrollY();
        RenderUtils.drawRoundedRect(this.modulex + 5.0f + this.pos.x + 55.0f, this.moduley + 17.0f + this.texty + 8.0f, 75.0f, 11.0f, 1.0f, new Color(59, 63, 72).getRGB(), 1.0f, this.textValue.getTextHovered() ? guiColor : new Color(85, 90, 96).getRGB());
        if (this.isHovered(mouseX, mouseY) && !this.textValue.getTextHovered()) {
            RenderUtils.drawRoundedRect(this.modulex + 5.0f + this.pos.x + 55.0f, this.moduley + 17.0f + this.texty + 8.0f, 75.0f, 11.0f, 1.0f, new Color(0, 0, 0, 0).getRGB(), 1.0f, guiColor);
        }
        Fonts.fontTahoma.drawString(this.textValue.getName(), this.modulex + 5.0f + this.pos.x + 4.0f, this.moduley + 17.0f + this.texty + 13.0f, new Color(200, 200, 200).getRGB());
        if (Fonts.fontTahoma.getStringWidth(this.textValue.getTextHovered() ? (this.textValue.get() + "_") : ((Value<String>)this.textValue).get()) > 70) {
            Fonts.fontTahoma.drawString(Fonts.fontTahoma.trimStringToWidth(this.textValue.getTextHovered() ? (this.textValue.get() + "_") : ((Value<String>)this.textValue).get(), 78, true), this.modulex + 5.0f + this.pos.x + 57.0f, this.moduley + 17.0f + this.texty + 13.0f, new Color(200, 200, 200).getRGB());
        }
        else if (this.textValue.get().isEmpty() && !this.textValue.getTextHovered()) {
            Fonts.fontTahoma.drawString("Type Here...", this.modulex + 5.0f + this.pos.x + 57.0f, this.moduley + 17.0f + this.texty + 13.0f, new Color(200, 200, 200).getRGB());
        }
        else {
            Fonts.fontTahoma.drawString(this.textValue.getTextHovered() ? (this.textValue.get() + "_") : ((Value<String>)this.textValue).get(), this.modulex + 5.0f + this.pos.x + 57.0f, this.moduley + 17.0f + this.texty + 13.0f, new Color(200, 200, 200).getRGB());
        }
    }
    
    public boolean isHovered(final int mouseX, final int mouseY) {
        return mouseX >= this.modulex + 5.0f + this.pos.x + 55.0f && mouseX <= this.modulex + 5.0f + this.pos.x + 55.0f + 75.0f && mouseY >= this.moduley + 17.0f + this.texty + 8.0f && mouseY <= this.moduley + 17.0f + this.texty + 8.0f + 11.0f;
    }
    
    @Override
    public void mouseClicked(final int mouseX, final int mouseY, final int mouseButton) {
        if (this.isHovered(mouseX, mouseY)) {
            this.textValue.setTextHovered(!this.textValue.getTextHovered());
        }
        else if (this.textValue.getTextHovered()) {
            this.textValue.setTextHovered(false);
        }
    }
    
    @Override
    public void keyTyped(final char typedChar, final int keyCode) {
        if (this.textValue.getTextHovered()) {
            if (keyCode == 1) {
                this.textValue.setTextHovered(false);
            }
            else if (keyCode != 14 && keyCode != 157 && keyCode != 29 && keyCode != 54 && keyCode != 42 && keyCode != 15 && keyCode != 58 && keyCode != 211 && keyCode != 199 && keyCode != 210 && keyCode != 200 && keyCode != 208 && keyCode != 205 && keyCode != 203 && keyCode != 56 && keyCode != 184 && keyCode != 197 && keyCode != 70 && keyCode != 207 && keyCode != 201 && keyCode != 209 && keyCode != 221 && keyCode != 59 && keyCode != 60 && keyCode != 61 && keyCode != 62 && keyCode != 63 && keyCode != 64 && keyCode != 65 && keyCode != 66 && keyCode != 67 && keyCode != 68 && keyCode != 87 && keyCode != 88) {
                this.textValue.append(typedChar);
            }
            if (((TextValue)this.setting).getTextHovered() && Keyboard.isKeyDown(14) && this.textValue.get().length() >= 1) {
                this.textValue.set(this.textValue.get().substring(0, this.textValue.get().length() - 1));
            }
        }
    }
    
    @Override
    public void mouseReleased(final int mouseX, final int mouseY, final int state) {
    }
}

File: src/main/java/net/ccbluex/liquidbounce/ui/client/clickgui/style/styles/nlclickgui/settings/TextSetting.kt
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/SkidderMC/FDPClient/
 */
package net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.settings

import net.ccbluex.liquidbounce.config.TextValue
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Downward
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.NeverloseGui
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.NlModule
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.RenderUtil
import net.ccbluex.liquidbounce.ui.font.Fonts
import net.minecraft.util.ChatAllowedCharacters
import org.lwjgl.input.Keyboard
import java.awt.Color
import kotlin.math.max

class TextSetting(setting: TextValue, moduleRender: NlModule) : Downward<TextValue>(setting, moduleRender) {

    private var editing = false
    private var buffer: String = setting.get()

    override fun draw(mouseX: Int, mouseY: Int) {
        val gui = NeverloseGui.getInstance()
        val mainx = gui.x
        val mainy = gui.y
        val textY = (y + getScrollY()).toInt()

        Fonts.Nl_16.drawString(
            setting.name,
            (mainx + 100 + x),
            (mainy + textY + 57).toFloat(),
            if (gui.light) Color(95, 95, 95).rgb else -1
        )

        val display = if (editing) buffer else setting.get()
        val stringWidth = Fonts.Nl_15.stringWidth(display) + 6

        RenderUtil.drawRoundedRect(
            (mainx + 170 + x),
            (mainy + textY + 54).toFloat(),
            max(80, stringWidth).toFloat(),
            14f,
            2f,
            if (gui.light) Color(255, 255, 255).rgb else Color(0, 5, 19).rgb,
            1f,
            Color(13, 24, 35).rgb
        )

        Fonts.Nl_15.drawString(
            display,
            mainx + 174 + x,
            (mainy + textY + 59).toFloat(),
            if (gui.light) Color(95, 95, 95).rgb else -1
        )

        if (!editing) {
            buffer = setting.get()
        } else {
            setting.set(buffer)
        }
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        val gui = NeverloseGui.getInstance()
        val boxX = (gui.x + 170 + x)
        val boxY = (gui.y + (y + getScrollY()).toInt() + 54).toFloat()

        val display = if (editing) buffer else setting.get()
        val stringWidth = Fonts.Nl_15.stringWidth(display) + 6
        val boxWidth = max(80, stringWidth).toFloat()

        if (mouseButton == 0) {
            editing = RenderUtil.isHovering(boxX, boxY, boxWidth, 14f, mouseX, mouseY)
            if (editing) {
                buffer = setting.get()
            }
        }
    }

    override fun mouseReleased(mouseX: Int, mouseY: Int, state: Int) {}

    override fun keyTyped(typedChar: Char, keyCode: Int) {
        if (!editing) return

        when (keyCode) {
            Keyboard.KEY_ESCAPE -> editing = false
            Keyboard.KEY_BACK -> if (buffer.isNotEmpty()) buffer = buffer.substring(0, buffer.length - 1)
            Keyboard.KEY_RETURN -> {
                setting.set(buffer)
                editing = false
            }

            else -> {
                if (ChatAllowedCharacters.isAllowedCharacter(typedChar)) {
                    buffer += typedChar
                }
            }
        }
    }
}